File: Swift/LeetCode/LeetCode/ThreeSum.swift
//
//  ThreeSum.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 03/12/2017.
//  Copyright © 2017 lizhaotailang. All rights reserved.
//
//  Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0?
//  Find all unique triplets in the array which gives the sum of zero.
//
//  Note: The solution set must not contain duplicate triplets.
//
//  For example, given array S = [-1, 0, 1, 2, -1, -4],
//
//  A solution set is:
//  [
//  [-1, 0, 1],
//  [-1, -1, 2]
//  ]
//
//  Accepted. See [ThreeSumTests](./LeetCodeTests/ThreeSumTests.swift) for test cases.
//

import Foundation

class ThreeSum {
    
    func threeSum(_ nums: [Int]) -> [[Int]] {
        let array = nums.sorted()
        
        var set = Set<Triple>.init()
        
        var first = 0
        while first < array.count - 2 {
            if array[first] > 0 {
                break
            }
            
            let target = 0 - array[first]
            var second = first + 1
            var third = array.count - 1
            
            while second < third {
                if array[second] + array[third] == target {
                    set.insert(Triple(array[first], array[second], array[third]))
                    while second < third && array[second] == array[second + 1] {
                        second += 1
                    }
                    while second < third && array[third] == array[third - 1] {
                        third -= 1
                    }
                    second += 1
                    third -= 1
                } else if array[second] + array[third] < target {
                    second += 1
                } else {
                    third -= 1
                }
            }
            
            first += 1
        }
        
        var lists = Array<Array<Int>>.init()
        set.forEach{
            lists.append([$0.a, $0.b, $0.c])
        }
        
        return lists
    }
 
    class Triple: Equatable, Hashable {
        
        var a: Int
        var b: Int
        var c: Int
        
        var hashValue: Int {
            var value = (-self.a) * 100
            value += abs(self.b) * 10
            value += c
            return value
        }
        
        init(_ a: Int, _ b: Int, _ c: Int) {
            self.a = a
            self.b = b
            self.c = c
        }
        
        static func ==(lhs: ThreeSum.Triple, rhs: ThreeSum.Triple) -> Bool {
            return lhs.a == rhs.a && lhs.b == rhs.b && lhs.c == rhs.c
        }
        
    }
}

File: Golang/ThreeSum/ThreeSum.go
package ThreeSum

//Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0?
//Find all unique triplets in the array which gives the sum of zero.
//
//Note: The solution set must not contain duplicate triplets.
//
//For example, given array S = [-1, 0, 1, 2, -1, -4],
//
//A solution set is:
//[
//[-1, 0, 1],
//[-1, -1, 2]
//]
//
//Accepted.

import (
	"sort"
)

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	set := make(map[[3]int]*Triple)

	for first := 0; first < len(nums)-2; first++ {
		if nums[first] > 0 {
			break
		}

		target := 0 - nums[first]
		second := first + 1
		third := len(nums) - 1

		for ; second < third; {
			if nums[second]+nums[third] == target {

				key := [3]int{nums[first], nums[second], nums[third]}

				if _, ok := set[key]; !ok {
					set[key] = &Triple{
						a: nums[first],
						b: nums[second],
						c: nums[third],
					}
				}

				for ; second < third && nums[second] == nums[second+1]; {
					second++
				}
				for ; second < third && nums[third] == nums[third-1]; {
					third--
				}
				second++
				third--
			} else if nums[second]+nums[third] < target {
				second++
			} else {
				third--
			}
		}
	}

	lists := make([][]int, 0)
	for _, v := range set {
		l := make([]int, 0)
		l = append(l, v.a)
		l = append(l, v.b)
		l = append(l, v.c)
		lists = append(lists, l)
	}

	return lists
}

type Triple struct {
	a int
	b int
	c int
}
